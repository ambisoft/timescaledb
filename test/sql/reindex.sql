CREATE TABLE reindex_test(time timestamp, temp float, PRIMARY KEY(time, temp));
CREATE UNIQUE INDEX reindex_test_time_unique_idx ON reindex_test(time);

-- create hypertable with three chunks
SELECT create_hypertable('reindex_test', 'time', chunk_time_interval => 2628000000000);

INSERT INTO reindex_test VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);

SELECT * FROM test.show_columns('reindex_test');
SELECT * FROM test.show_subtables('reindex_test');

-- show reindexing
REINDEX (VERBOSE) TABLE reindex_test;

\set ON_ERROR_STOP 0
-- this one currently doesn't recurse to chunks and instead gives an
-- error
REINDEX (VERBOSE) INDEX reindex_test_time_unique_idx;
\set ON_ERROR_STOP 1

-- show reindexing on a normal table
CREATE TABLE reindex_norm(time timestamp, temp float);
CREATE UNIQUE INDEX reindex_norm_time_unique_idx ON reindex_norm(time);

INSERT INTO reindex_norm VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);

REINDEX (VERBOSE) TABLE reindex_norm;
REINDEX (VERBOSE) INDEX reindex_norm_time_unique_idx;

SELECT * FROM test.show_constraintsp('_timescaledb_internal.%');
SELECT * FROM reindex_norm;

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');

SELECT indexrelid AS original_indexoid FROM pg_index WHERE indexrelid = '_timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx'::regclass
\gset

SELECT reindex('reindex_test', verbose=>true);

SELECT indexrelid AS reindex_indexoid FROM pg_index WHERE indexrelid = '_timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx'::regclass
\gset

SELECT :original_indexoid = :reindex_indexoid;

SELECT reindex('reindex_test', verbose=>true, recreate=>true);

SELECT indexrelid AS remake_indexoid FROM pg_index WHERE indexrelid = '_timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx'::regclass
\gset

SELECT :original_indexoid = :remake_indexoid;

--reindex only a particular index
SELECT reindex('reindex_test_pkey', verbose=>true, recreate=>true);

--from_time
SELECT reindex('reindex_test_pkey', time_column=>'time', from_time=>TIMESTAMP '2017-06-20T12:00:00', verbose=>true, recreate=>true);

--to_time (note awkwardness with having to declare a NULL from_time)
SELECT reindex('reindex_test_pkey', 
    time_column=>'time', from_time=>NULL::timestamp, to_time=>TIMESTAMP '2017-02-18 20:00:00',
    verbose=>true, recreate=>true);

--both from and to time
SELECT reindex('reindex_test_pkey', 
    time_column=>'time', from_time=>'2017-01-19 10:00:00'::timestamp, to_time=>TIMESTAMP '2017-02-18 20:00:00', 
    verbose=>true, recreate=>true);

--empty set
SELECT reindex('reindex_test_pkey', 
    time_column=>'time', from_time=>'2017-01-19 06:00:00'::timestamp, to_time=>TIMESTAMP '2017-02-18 15:00:00', 
    verbose=>true, recreate=>true);

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
