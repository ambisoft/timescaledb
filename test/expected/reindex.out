CREATE TABLE reindex_test(time timestamp, temp float, PRIMARY KEY(time, temp));
CREATE UNIQUE INDEX reindex_test_time_unique_idx ON reindex_test(time);
-- create hypertable with three chunks
SELECT create_hypertable('reindex_test', 'time', chunk_time_interval => 2628000000000);
 create_hypertable 
-------------------
 
(1 row)

INSERT INTO reindex_test VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);
SELECT * FROM test.show_columns('reindex_test');
 Column |            Type             | Nullable 
--------+-----------------------------+----------
 time   | timestamp without time zone | t
 temp   | double precision            | t
(2 rows)

SELECT * FROM test.show_subtables('reindex_test');
                 Child                  
----------------------------------------
 _timescaledb_internal._hyper_1_1_chunk
 _timescaledb_internal._hyper_1_2_chunk
 _timescaledb_internal._hyper_1_3_chunk
 _timescaledb_internal._hyper_1_4_chunk
 _timescaledb_internal._hyper_1_5_chunk
(5 rows)

-- show reindexing
REINDEX (VERBOSE) TABLE reindex_test;
INFO:  index "_hyper_1_1_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "1_1_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_2_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "2_2_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_3_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "3_3_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_4_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "4_4_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_5_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "5_5_reindex_test_pkey" was reindexed
\set ON_ERROR_STOP 0
-- this one currently doesn't recurse to chunks and instead gives an
-- error
REINDEX (VERBOSE) INDEX reindex_test_time_unique_idx;
ERROR:  Reindexing of a specific index on a hypertable is currently unsupported.
\set ON_ERROR_STOP 1
-- show reindexing on a normal table
CREATE TABLE reindex_norm(time timestamp, temp float);
CREATE UNIQUE INDEX reindex_norm_time_unique_idx ON reindex_norm(time);
INSERT INTO reindex_norm VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);
REINDEX (VERBOSE) TABLE reindex_norm;
INFO:  index "reindex_norm_time_unique_idx" was reindexed
REINDEX (VERBOSE) INDEX reindex_norm_time_unique_idx;
INFO:  index "reindex_norm_time_unique_idx" was reindexed
SELECT * FROM test.show_constraintsp('_timescaledb_internal.%');
                 Table                  |      Constraint       | Type |   Columns   |                     Index                     |                                                                     Expr                                                                     
----------------------------------------+-----------------------+------+-------------+-----------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------
 _timescaledb_internal._hyper_1_1_chunk | 1_1_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."1_1_reindex_test_pkey" | 
 _timescaledb_internal._hyper_1_1_chunk | constraint_1          | c    | {time}      | -                                             | (("time" >= 'Thu Jan 19 10:00:00 2017'::timestamp without time zone) AND ("time" < 'Sat Feb 18 20:00:00 2017'::timestamp without time zone))
 _timescaledb_internal._hyper_1_2_chunk | 2_2_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."2_2_reindex_test_pkey" | 
 _timescaledb_internal._hyper_1_2_chunk | constraint_2          | c    | {time}      | -                                             | (("time" >= 'Tue Mar 21 06:00:00 2017'::timestamp without time zone) AND ("time" < 'Thu Apr 20 16:00:00 2017'::timestamp without time zone))
 _timescaledb_internal._hyper_1_3_chunk | 3_3_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."3_3_reindex_test_pkey" | 
 _timescaledb_internal._hyper_1_3_chunk | constraint_3          | c    | {time}      | -                                             | (("time" >= 'Thu Apr 20 16:00:00 2017'::timestamp without time zone) AND ("time" < 'Sun May 21 02:00:00 2017'::timestamp without time zone))
 _timescaledb_internal._hyper_1_4_chunk | 4_4_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."4_4_reindex_test_pkey" | 
 _timescaledb_internal._hyper_1_4_chunk | constraint_4          | c    | {time}      | -                                             | (("time" >= 'Sun May 21 02:00:00 2017'::timestamp without time zone) AND ("time" < 'Tue Jun 20 12:00:00 2017'::timestamp without time zone))
 _timescaledb_internal._hyper_1_5_chunk | 5_5_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."5_5_reindex_test_pkey" | 
 _timescaledb_internal._hyper_1_5_chunk | constraint_5          | c    | {time}      | -                                             | (("time" >= 'Tue Jun 20 12:00:00 2017'::timestamp without time zone) AND ("time" < 'Thu Jul 20 22:00:00 2017'::timestamp without time zone))
(10 rows)

SELECT * FROM reindex_norm;
           time           | temp 
--------------------------+------
 Fri Jan 20 09:00:01 2017 | 17.5
 Sat Jan 21 09:00:01 2017 | 19.1
 Thu Apr 20 09:00:01 2017 | 89.5
 Fri Apr 21 09:00:01 2017 | 17.1
 Tue Jun 20 09:00:01 2017 | 18.5
 Wed Jun 21 09:00:01 2017 |   11
(6 rows)

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
                                Index                                |   Columns   | Unique | Primary | Exclusion | Tablespace 
---------------------------------------------------------------------+-------------+--------+---------+-----------+------------
 _timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx | {time}      | t      | f       | f         | 
 _timescaledb_internal."1_1_reindex_test_pkey"                       | {time,temp} | t      | t       | f         | 
(2 rows)

SELECT indexrelid AS original_indexoid FROM pg_index WHERE indexrelid = '_timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx'::regclass
\gset
SELECT reindex('reindex_test', verbose=>true);
INFO:  index "_hyper_1_1_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "1_1_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_2_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "2_2_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_3_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "3_3_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_4_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "4_4_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_5_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "5_5_reindex_test_pkey" was reindexed
 reindex 
---------
 
(1 row)

SELECT indexrelid AS reindex_indexoid FROM pg_index WHERE indexrelid = '_timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx'::regclass
\gset
SELECT :original_indexoid = :reindex_indexoid;
 ?column? 
----------
 t
(1 row)

SELECT reindex('reindex_test', verbose=>true, recreate=>true);
INFO:  a new index was created for index "_hyper_1_1_chunk_reindex_test_time_unique_idx"
INFO:  a new index was created for index "1_1_reindex_test_pkey"
INFO:  a new index was created for index "_hyper_1_2_chunk_reindex_test_time_unique_idx"
INFO:  a new index was created for index "2_2_reindex_test_pkey"
INFO:  a new index was created for index "_hyper_1_3_chunk_reindex_test_time_unique_idx"
INFO:  a new index was created for index "3_3_reindex_test_pkey"
INFO:  a new index was created for index "_hyper_1_4_chunk_reindex_test_time_unique_idx"
INFO:  a new index was created for index "4_4_reindex_test_pkey"
INFO:  a new index was created for index "_hyper_1_5_chunk_reindex_test_time_unique_idx"
INFO:  a new index was created for index "5_5_reindex_test_pkey"
INFO:  index "_hyper_1_1_chunk_reindex_test_time_unique_idx" was renamed to use the new index
INFO:  index "1_1_reindex_test_pkey" was renamed to use the new index
INFO:  index "_hyper_1_2_chunk_reindex_test_time_unique_idx" was renamed to use the new index
INFO:  index "2_2_reindex_test_pkey" was renamed to use the new index
INFO:  index "_hyper_1_3_chunk_reindex_test_time_unique_idx" was renamed to use the new index
INFO:  index "3_3_reindex_test_pkey" was renamed to use the new index
INFO:  index "_hyper_1_4_chunk_reindex_test_time_unique_idx" was renamed to use the new index
INFO:  index "4_4_reindex_test_pkey" was renamed to use the new index
INFO:  index "_hyper_1_5_chunk_reindex_test_time_unique_idx" was renamed to use the new index
INFO:  index "5_5_reindex_test_pkey" was renamed to use the new index
 reindex 
---------
 
(1 row)

SELECT indexrelid AS remake_indexoid FROM pg_index WHERE indexrelid = '_timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx'::regclass
\gset
SELECT :original_indexoid = :remake_indexoid;
 ?column? 
----------
 f
(1 row)

--reindex only a particular index
SELECT reindex('reindex_test_pkey', verbose=>true, recreate=>true);
INFO:  a new index was created for index "1_1_reindex_test_pkey"
INFO:  a new index was created for index "2_2_reindex_test_pkey"
INFO:  a new index was created for index "3_3_reindex_test_pkey"
INFO:  a new index was created for index "4_4_reindex_test_pkey"
INFO:  a new index was created for index "5_5_reindex_test_pkey"
INFO:  index "1_1_reindex_test_pkey" was renamed to use the new index
INFO:  index "2_2_reindex_test_pkey" was renamed to use the new index
INFO:  index "3_3_reindex_test_pkey" was renamed to use the new index
INFO:  index "4_4_reindex_test_pkey" was renamed to use the new index
INFO:  index "5_5_reindex_test_pkey" was renamed to use the new index
 reindex 
---------
 
(1 row)

--from_time
SELECT reindex('reindex_test_pkey', time_column=>'time', from_time=>TIMESTAMP '2017-06-20T12:00:00', verbose=>true, recreate=>true);
INFO:  a new index was created for index "5_5_reindex_test_pkey"
INFO:  index "5_5_reindex_test_pkey" was renamed to use the new index
 reindex 
---------
 
(1 row)

--to_time (note awkwardness with having to declare a NULL from_time)
SELECT reindex('reindex_test_pkey', 
    time_column=>'time', from_time=>NULL::timestamp, to_time=>TIMESTAMP '2017-02-18 20:00:00',
    verbose=>true, recreate=>true);
INFO:  a new index was created for index "1_1_reindex_test_pkey"
INFO:  index "1_1_reindex_test_pkey" was renamed to use the new index
 reindex 
---------
 
(1 row)

--both from and to time
SELECT reindex('reindex_test_pkey', 
    time_column=>'time', from_time=>'2017-01-19 10:00:00'::timestamp, to_time=>TIMESTAMP '2017-02-18 20:00:00', 
    verbose=>true, recreate=>true);
INFO:  a new index was created for index "1_1_reindex_test_pkey"
INFO:  index "1_1_reindex_test_pkey" was renamed to use the new index
 reindex 
---------
 
(1 row)

--empty set
SELECT reindex('reindex_test_pkey', 
    time_column=>'time', from_time=>'2017-01-19 06:00:00'::timestamp, to_time=>TIMESTAMP '2017-02-18 15:00:00', 
    verbose=>true, recreate=>true);
 reindex 
---------
 
(1 row)

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
                                Index                                |   Columns   | Unique | Primary | Exclusion | Tablespace 
---------------------------------------------------------------------+-------------+--------+---------+-----------+------------
 _timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx | {time}      | t      | f       | f         | 
 _timescaledb_internal."1_1_reindex_test_pkey"                       | {time,temp} | t      | t       | f         | 
(2 rows)

